From: Jan Kara <jack@suse.cz>
Subject: Make ext4 mounts read-only unless proper config option is specified
References: fate#311111
Patch-mainline: never

All ext4 mounts should be read-only to allow migration. We keep possibility
of read-write mounts when CONFIG_EXT4_FS_RW is defined so that we can easily
build unsupporeted KMP with full ext4 support.

Signed-off-by: Jan Kara <jack@suse.cz>

diff -rupX /crypted/home/jack/.kerndiffexclude linux-2.6.32-SLE11-SP2/fs/ext4/super.c linux-2.6.32-SLE11-SP2-1-ext4rw/fs/ext4/super.c
--- linux-2.6.32-SLE11-SP2/fs/ext4/super.c	2011-06-15 01:08:43.450202556 +0200
+++ linux-2.6.32-SLE11-SP2-1-ext4rw/fs/ext4/super.c	2011-06-15 23:46:38.783042601 +0200
@@ -2423,6 +2423,10 @@ static int ext4_fill_super(struct super_
 		goto cantfind_ext4;
 	sbi->s_kbytes_written = le64_to_cpu(es->s_kbytes_written);
 
+#ifndef CONFIG_EXT4_FS_RW
+	sb->s_flags |= MS_RDONLY;
+	ext4_msg(sb, KERN_INFO, "ext4 is supported in read-only mode only");
+#endif
 	/* Set defaults before we parse the mount options */
 	def_mount_opts = le32_to_cpu(es->s_default_mount_opts);
 	if (def_mount_opts & EXT4_DEFM_DEBUG)
@@ -3510,6 +3514,14 @@ static int ext4_remount(struct super_blo
 	if (sbi->s_journal && sbi->s_journal->j_task->io_context)
 		journal_ioprio = sbi->s_journal->j_task->io_context->ioprio;
 
+#ifndef CONFIG_EXT$_FS_RW
+	if (!(*flags & MS_RDONLY)) {
+		*flags |= MS_RDONLY;
+		ext4_msg(sb, KERN_INFO,
+			 "ext4 is supported in read-only mode only");
+	}
+#endif
+
 	/*
 	 * Allow the "check" option to be passed as a remount option.
 	 */
